File: sections/67_persistence_rollback_time/problems/a_branching_multiset/solution.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node {
  int left = 0, right = 0, count = 0;
};
vector<Node> tree(1);
int update(int old, int lo, int hi, int position, int delta) {
  int node = tree.size();
  tree.push_back(tree[old]);
  tree[node].count += delta;
  if (lo != hi) {
    int mid = (lo + hi) / 2;
    if (position <= mid)
      tree[node].left = update(tree[old].left, lo, mid, position, delta);
    else
      tree[node].right = update(tree[old].right, mid + 1, hi, position, delta);
  }
  return node;
}
int prefix(int node, int lo, int hi, int x) {
  if (!node || x < lo) return 0;
  if (hi <= x) return tree[node].count;
  int mid = (lo + hi) / 2;
  return prefix(tree[node].left, lo, mid, x) + prefix(tree[node].right, mid + 1, hi, x);
}
int kth(int node, int lo, int hi, int k) {
  if (lo == hi) return lo;
  int mid = (lo + hi) / 2, left_count = tree[tree[node].left].count;
  if (k <= left_count) return kth(tree[node].left, lo, mid, k);
  return kth(tree[node].right, mid + 1, hi, k - left_count);
}
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int maximum, q;
  if (!(cin >> maximum >> q)) return 0;
  vector<int> roots(1, 0);
  while (q--) {
    char type;
    int version, x;
    cin >> type >> version >> x;
    if (type == 'I' || type == 'E')
      roots.push_back(update(roots[version], 1, maximum, x, type == 'I' ? 1 : -1));
    else if (type == 'K')
      cout << kth(roots[version], 1, maximum, x) << '\n';
    else
      cout << prefix(roots[version], 1, maximum, x) << '\n';
  }
}

File: sections/67_persistence_rollback_time/problems/a_branching_multiset/solution.py
import sys


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    maximum, q = map(int, data[:2])
    index = 2
    left = [0]
    right = [0]
    count = [0]
    roots = [0]

    def update(old, lo, hi, pos, delta):
        node = len(count)
        left.append(left[old])
        right.append(right[old])
        count.append(count[old] + delta)
        if lo != hi:
            mid = (lo + hi) // 2
            if pos <= mid:
                left[node] = update(left[old], lo, mid, pos, delta)
            else:
                right[node] = update(right[old], mid + 1, hi, pos, delta)
        return node

    def prefix(node, lo, hi, x):
        if not node or x < lo:
            return 0
        if hi <= x:
            return count[node]
        mid = (lo + hi) // 2
        return prefix(left[node], lo, mid, x) + prefix(right[node], mid + 1, hi, x)

    def kth(node, lo, hi, k):
        while lo != hi:
            mid = (lo + hi) // 2
            amount = count[left[node]]
            if k <= amount:
                node = left[node]
                hi = mid
            else:
                k -= amount
                node = right[node]
                lo = mid + 1
        return lo

    output = []
    for _ in range(q):
        kind = data[index].decode()
        version = int(data[index + 1])
        x = int(data[index + 2])
        index += 3
        if kind in "IE":
            roots.append(
                update(roots[version], 1, maximum, x, 1 if kind == "I" else -1)
            )
        elif kind == "K":
            output.append(str(kth(roots[version], 1, maximum, x)))
        else:
            output.append(str(prefix(roots[version], 1, maximum, x)))
    print("\n".join(output))


if __name__ == "__main__":
    main()
